### ISSUE TITLE
Onnxruntime-gpu always runs on CPU and never on GPU, [Roberta]

### ISSUE THREAD

**Issue by Oxi84** [microsoft/onnxruntime#9292]
The code: it first loads Roberta base, prepares input and convert model to onnx and the load and run it,

    import torch
    import time
    from transformers import RobertaTokenizerFast,RobertaForMaskedLM,BertForMaskedLM
    the_model_rb = "roberta-base" 
    tokenizer = RobertaTokenizerFast.from_pretrained(the_model_rb,cache_dir="/root/Desktop/python/onxx/m/")
    model = RobertaForMaskedLM.from_pretrained(the_model_rb,cache_dir="/root/Desktop/python/onxx/m/")
    model.eval() 
    model.to('cuda')
    print("the model is loaded")
    
    text = "<s> Cuda does not works. </s>"
    tokenized_text = tokenizer.tokenize(text)
    print("tokenized_text",tokenized_text)
    maskedindex = 3
    tokenized_text[maskedindex] = "<mask>"
    indexed_tokens = tokenizer.convert_tokens_to_ids(tokenized_text)
    attentions_mask = [1 for x in range(0,len(tokenized_text))]            
        
    input_mask = torch.tensor([attentions_mask,attentions_mask]).to("cuda")
    input_ids = torch.tensor([indexed_tokens,indexed_tokens]).to("cuda") 
    
    dynamic_axes_d1={'input_ids': {0: 'batch',1: 'sequence'},'input_mask': {0: 'batch', 1: 'sequence'}, 'output': {0: 'batch', 1: 'sequence'}}
        
    torch.onnx.export(model, (input_ids,input_mask), "/root/Desktop/onnx/robertabase.onnx",input_names = ["input_ids", "input_mask"],
        output_names = ["output"], verbose=False, opset_version=11, do_constant_folding=True,dynamic_axes=dynamic_axes_d1)
    print("we have converted the model dynamic")
    
    
    #load the model:           
    import onnxruntime  
    print(onnxruntime.get_device())  
    ONNX_PROVIDERS = ["CUDAExecutionProvider"]
    model_dir = "/root/Desktop/onnx/robertabase.onnx"
    the_session = onnxruntime.InferenceSession(model_dir,providers=ONNX_PROVIDERS)   
    print(the_session.get_providers())     

    import numpy
    def to_numpy(tensor):
        return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()    
    
    #excute the model
    beg = time.time()
    for i in range(10):
        ort_inputs = {the_session.get_inputs()[0].name: input_ids.detach().cpu().numpy(), the_session.get_inputs()[1].name: input_mask.detach().cpu().numpy()}
        ort_outs = the_session.run(["output"], ort_inputs)   
        torch_onnx_output = torch.tensor(ort_outs[0])
    print("onxx took",-beg + time.time())

It shows that onnxruntime.get_device() is GPU, but then the_session.get_providers() is only Cpu.provider and never GPU provider.

Expected it would be that almost everything is done on cuda, so the provider needs to be Cuda.

It is on Ubuntu 18.04 , latest version of onnxruntime-gpu at the moment, i have just installed it, i think pytoch 1.8 and cuda 10.2 on a anaconda environment.

**Comment by jcwchen** [comment]
No worries. Then please check your CUDA and CuDNN versions. They must be compatible to each other. Here is the link which might be helpful: https://onnxruntime.ai/docs/execution-providers/CUDA-ExecutionProvider.html For ONNXRuntime 1.9.0 package, I think it was built by CUDA 11.4. Ideally it should support 11.x with the minimum required driver version... I will confirm with the team which CUDA version the minimum required is for ORT 1.9.0 and let you know. Thank you!

BTW, if you are looking for a way to build ORT from source with CUDA EP, here is the tutorial: https://onnxruntime.ai/docs/build/eps.html#cuda.

**Comment by Oxi84** [comment]
> 
> 
> Hi @Oxi84, Please make sure you have only onnxruntime-gpu installed. Have you also installed onnxruntime (CPU-only)? If so, could you please `pip uninstall -y onnxruntime && pip uninstall -y onnxruntime-gpu && pip install onnxruntime-gpu` and try again? Thank you.

I did install only onnxruntime-gpu and when i do: print(onnxruntime.version) i get 1.9.0
print(onnxruntime.get_device()) gives me GPU
print(the_session.get_providers()) gives me CPUexecution

All is installed in an anacodan envinronment, so every anaconda environment has its own version of cuda and this one has cuda 10.2 not sure which CUDNN.

Do i need to install another version of onnxruntime that is compatible with cuda 10.2?

Also it is possible that it onxxruntime tries to use system's Cuda and not the one from the conda envinronment, even for example pytorch detects and uses the right version of cuda.

**Comment by Oxi84** [comment]
> 
> 
> Hi, @jcwchen, @Oxi84 . I face same problem, and I found this Bug is fixed (for the issue that fails GPU execution if the executable is on the path that contained the unicode characters) in 1.9.1, [https://github.com/microsoft/onnxruntime/releases/tag/v1.9.1](url)
> 
> but I don't know how to install 1.9.1 from the source code in the link. May you give me some guidance? Thanks a lot

Do you get GPU as device  - print(onnxruntime.get_device()) gives me GPU

**Comment by Oxi84** [comment]
I cannot install older version using line  pip install onnxruntime-gpu-1.8.1, is there any other way as i found cuda 10.2 was faster than 11 and 11.2 last time i checked.

**Comment by Oxi84** [comment]
What speedup could I expect from ONNX for Roberta or Bert if batch size is around 10 and sequence is around 60 on average?

Anyways, will try a clean system install. I use anaconda and then i install cuda only in virtual environments, i guess there is no way to use it as a system cuda when declaring it in some environment variable?

I use the optimiser in the exact way you showed, I get around 10+ percent speedup from it on CPU. So totally it is around 33 percent of speedup. 4 sec without and 3 sec with ONNX for batch size 10 on CPU which is the fastest.

**Comment by Oxi84** [comment]
I made it to work using cuda 11, and even the onxx model is only 600 mb, onxx uses around 2400 mb of memory. And pytorch uses around 1200 mb of memory, so the memory usage is around 2x more.

And ONXX should use less memory, as far as i read right.

I used the same code as above to convert the model.

**Comment by JohnHerry** [comment]
> I made it to work using cuda 11, and even the onxx model is only 600 mb, onxx uses around 2400 mb of memory. And pytorch uses around 1200 mb of memory, so the memory usage is around 2x more.
> 
> And ONXX should use less memory, as far as i read right.
> 
> I used the same code as above to convert the model.

I am under a cuda12.1 and cudnn8.9; When I install the cuda12 compatible version of onnxruntime, like   `pip install onnxruntime-gpu==1.17.1 --extra-index-url https://aiinfra.pkgs.visualstudio.com/PublicPackages/_packaging/onnxruntime-cuda-12/pypi/simple/`  
it will report Duplicate GPU error. because I use onnxruntime-gpu to extract input feature for pytorch model training with torch.distributed.    error eg: `Duplicate GPU detected: rank 3 and rank 0 both on cuda device  18000   ncclInvalidUsage: This usually reflects invalid usage of NCCL library`   
while when I use the directly installed package `pip install onnxruntime-gpu==1.17.1`    it report  `provider_bridge_ort.cc:1209 onnxruntime::Provider& onnxruntime::ProviderLibrary::Get()  Failed to load library libonnxruntime_providers_cuda.so with error: libcublasLt.so.11: cannot open shared object file: No such file or directory`    
My sysmtem is Ubuntu==22.04 , cuda==12.1   python==3.10   cudnn==8.9   but defaut, there is a libcublasLt.so.12 in the lib dir, no libcublasLt.so.11
